staging/src/k8s.io/code-generator/cmd/generic-conversion-gen/generators/unsafe_conversion_arbitrator.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generators
+
+import (
+	"k8s.io/gengo/types"
+)
+
+// an UnsafeConversionArbitrator decides whether conversions can be done using unsafe casts.
+type UnsafeConversionArbitrator struct {
+	processedPairs           map[ConversionPair]unsafeConversionDecision
+	manualConversionsTracker *ManualConversionsTracker
+	functionTagName          string
+}
+
+type unsafeConversionDecision int
+
+const (
+	// notPossibleOneWay means that we've determined that the unsafe conversion from x to y
+	// is not possible, but that decision doesn't hold for conversions from y to x.
+	// That happens when there is a non-copy manual conversion function defined from x to y.
+	notPossibleOneWay unsafeConversionDecision = iota
```

Comment:
Not asking for a change, but in the future I strongly recommend keeping refactors and behavior changes 100% separated; it makes it so much easier to review (and also, I think, to not break anything while making the change).